components/service/glean/src/main/java/mozilla/components/service/glean/utils/FreshBootDetector.kt

```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.utils
+
+import android.content.Context
+import android.os.Build
+import android.provider.Settings
+import android.support.annotation.VisibleForTesting
+import java.io.File
+import java.io.IOException
+import java.util.UUID
+
+internal object FreshBootDetector {
+    // cache the value of the first check so we can safely run it multiple times
+    @VisibleForTesting
+    var cacheIsFreshBoot: Boolean? = null
+
+    /**
+     * Gets the boot id, which is a value that changes every time the
+     * device is rebooted.
+     *
+     * For Android Oreo (SDK 24) and later, we can use the [BOOT_COUNT] global
+     * setting.
+     *
+     * For earlier versions of Android, we can use the Linux kernel's
+     * `/proc/sys/kernel/random/boot_id`.
+     *
+     * Adapted from https://github.com/instacart/truetime-android/pull/93
+     *
+     * @param context An application context
+     */
+    internal fun getBootId(context: Context): String {
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O) {
+            try {
+                return Settings.Global.getInt(
+                    context.getContentResolver(),
+                    Settings.Global.BOOT_COUNT
+                ).toString()
+            } catch (e: Settings.SettingNotFoundException) {
+                // pass through to old method
+            }
+        }
+
+        try {
+            val file = File("/proc/sys/kernel/random/boot_id")
+
+            val bootId = StringBuilder()
+
+            file.forEachLine {
+                bootId.append(it)
+            }
+
+            return bootId.toString()
+        } catch (e: IOException) {
+            // pass through to fallback method
+        }
+
+        // If neither of the above methods works, just return
+        // a random id, so it will appear that we are always at
+        // a fresh reboot.
+        return UUID.randomUUID().toString()
```

Comment:
It would be interesting to know how often that happens. This would be much simpler if it would be possible to apply this fallback to all devices. :)

Reply:
Yes.  The motivation here is that the app may be shut down and restarted many times during the lifetime of a custom ping containing events, and we don't want to be sending extra pings on every process start.  If the events from the last run were during the same boot of the device, their timestamps will be compatible.  If they came from a previous boot, we have to send immediately and "start over" (because relying on realtime clocks to close the gap seemed ever more complex)...

It seems like the standard alternative is to listen to the `BOOT_COMPLETE` intent.  I thought that would be more problematic since it would start the app (at least in the background) on boot.  Might that still be preferred, just because it's a more well-trodden path?

Reply:
> It seems like the standard alternative is to listen to the BOOT_COMPLETE intent. I thought that would be more problematic since it would start the app (at least in the background) on boot.

That sounds okay-ish since it would only start your broadcast receiver (and initialize the application object). It also feels more like using a defined behavior than accessing something that is not really guaranteed to be accessible?

Anyhow, if we land as-is then let's test this on the different API levels and let's see how it behaves in the reference browser. :)

Reply:
I'll probably try out the `BOOT_COMPLETE` intent approach to see how that shakes out, so we have a point of comparison.